napi_basics.md

```diff
@@ -0,0 +1,139 @@
+- error_message: UTF8-encoded string containing a VM-neutral description of 
+  the error
+- engine_reserved: Reserved for VM-specific error details. This is currently
+  not implemented for any VM.
+- engine_error_code: VM-specific error code
+- error_code: the N-API status code that originated with the current error
+
+### Additional Information
+
+See the Error Handling section (<PLACEHOLDER>)
+
+### *napi_env*
+
+`napi_env` is used to represent a context that the underlying N-API 
+implementation can use to persist VM-specific state. This structure is passed
+to native functions when they're invoked, and it must be passed back when 
+making N-API calls. Specifically, the same `napi_env` that was passed in when
+the initial native function was called should be passed to any subsequent 
+nested N-API calls. Caching the `napi_env` for the purpose of general reuse is
+not allowed.
+
+### *napi_value*
+This is an opaque pointer that is used to represent a JavaScript value.
+
+## N-API Memory Management types
+
+### *napi_handle_scope*
+This is an abstraction used to control and modify the lifetime of objects 
+created within a particular scope. In general, N-API values are created within
+the context of a handle scope. When a native method is called from 
+JavaScript, a default handle scope will exist- if the user does not explicitly
+create a new handle scope, N-API values will be created in the default handle
+scope. For any invocations of code outside the execution of a native method 
+(for instance, during a libuv callback invocation), the module is required to 
+create a scope before invoking any functions that can result in the creation
+of JavaScript objects.
+
+Handle scopes are created using `napi_open_handle_scope` and are destroyed 
+using `napi_close_handle_scope`. Calling close can indicate to the GC that all
+napi_values created during the lifetime of the handle scope are no longer 
+referenced from the current stack frame.
+
+Consider creating a `napi_handle_scope` per function if you create any handle 
+other than just the return value. Otherwise, the handle created will leak to 
+the parent handle scope which can lead to higher than expected (and necessary)
+memory usage.
+
+### *napi_escapable_handle_scope*
+Escapable Handle Scopes are a special type of handle scope that are used to 
+return values created within a particular handle scope to a parent scope.
+
+### *napi_ref*
+This is the abstraction to use to reference a `napi_value`. This allows for 
+users to manage the lifetimes of JavaScript values, including defining their 
+minimum lifetimes explicitly.
+
+## N-API Callback types
+
+### *napi_callback_info*
+Opaque datatype that is passed to a callback function. It can be used for two
+purposes:
+- Get additional information about the context in which the callback was 
+  invoked
+- Set the return value of the callback
+
+### *napi_callback*
+Function pointer type for N-API APIs that call back into user-provided native 
```

Comment:
Instead of the first sentence I would write "Function pointer type for user-provided native functions which are to be exposed to JavaScript via N-API."